### PR TITLE
Update request and notification definitions to LSP 3.17

### PR DESCRIPTION
**Pull request by ahoppen**
This updates the definitions of the requests and notifications to match version 3.17 of the LSP spec

rdar://83833293

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/LanguageServerProtocol/Requests/SelectionRangeRequest.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+public struct SelectionRangeRequest: TextDocumentRequest {
+  public static var method: String = "textDocument/selectionRange"
+  public typealias Response = SelectionRange
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/Requests/SelectionRangeRequest.swift:15]: I think this should be `[SelectionRange]`

File: Sources/LanguageServerProtocol/Requests/SelectionRangeRequest.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }
+  }
+
+  enum CodingKeys: String, CodingKey {
+    case range
+    case _parent = "parent"
+  }
+
+  /// The range of this selection range.
+  public var range: Range<Position>
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/Requests/SelectionRangeRequest.swift:53]: Needs `@CustomCodable<PositionRange>`

File: Sources/LanguageServerProtocol/Requests/SignatureHelpRequest.swift [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+public struct ParameterInformation: Codable, Hashable {
+  public enum Label: Codable, Hashable {
+    case string(String)
+    case offsets(start: Int, end: Int)
+
+    public init(from decoder: Decoder) throws {
+      if let string = try? String(from: decoder) {
+        self = .string(string)
+      } else if let offsets = try? Array<Int>(from: decoder), offsets.count == 2 {
+        self = .offsets(start: offsets[0], end: offsets[2])
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/Requests/SignatureHelpRequest.swift:162]: Should probably be `offsets[1]`?

File: Sources/LanguageServerProtocol/Requests/SignatureHelpRequest.swift [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+/// Represents a parameter of a callable-signature. A parameter can
+/// have a label and a doc-comment.
+public struct ParameterInformation: Codable, Hashable {
+  public enum Label: Codable, Hashable {
+    case string(String)
+    case offsets(start: Int, end: Int)
+
+    public init(from decoder: Decoder) throws {
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/Requests/SignatureHelpRequest.swift:158]: Deserves a test in CodingTests.swift

**Comment by benlangmuir** [comment]
All the new request/notification types need to be added to the builtin message lists in Messages.swift or they won't decode correctly.

File: Sources/LanguageServerProtocol/SupportTypes/PositionEncoding.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+/// A set of predefined position encoding kinds.
+public struct PositionEncodingKind: RawRepresentable, Codable, Hashable {
+  public var rawValue: String
+
+  public init(rawValue: String) {
+    self.rawValue = rawValue
+  }
+
+  /// Character offsets count UTF-8 code units.
+  public static let utf8: PositionEncodingKind = PositionEncodingKind(rawValue: "utf-8")
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/LanguageServerProtocol/SupportTypes/PositionEncoding.swift:22]: Oh, it totally will be. I don’t even know how we would best implement this.

I’m viewing `LangaugeServerProtocol` as a general swift library for all the LSP types by now that isn’t necessarily tied to SourceKit-LSP. So I think it’s fine if we have this definition in the `LanguageServerProtocol` module but simply aren’t handling it in SourceKit-LSP.

File: Sources/LanguageServerProtocol/SupportTypes/WorkspaceEdit.swift [hunk 6]
```diff
@@ -18,10 +18,20 @@ public struct WorkspaceEdit: Hashable, ResponseType {
 
   public var documentChanges: [WorkspaceEditDocumentChange]?
 
+  /// A map of change annotations that can be referenced in
+  /// `AnnotatedTextEdit`s or create, rename and delete file / folder
+  /// operations.
+  ///
+  /// Whether clients honor this property depends on the client capability
+  /// `workspace.changeAnnotationSupport`.
+  public var changeAnnotations: [[ChangeAnnotationIdentifier]: ChangeAnnotation]?
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/LanguageServerProtocol/SupportTypes/WorkspaceEdit.swift:27]: Oh, you’re totally right. Thanks.

File: Sources/SourceKitLSP/Clang/ClangLanguageServer.swift [hunk 7]
```diff
@@ -431,7 +431,7 @@ extension ClangLanguageServerShim {
     // with `forceRebuild` set in case any missing header files have been added.
     // This works well for us as the moment since clangd ignores the document version.
     let note = DidChangeTextDocumentNotification(
-      textDocument: VersionedTextDocumentIdentifier(uri, version: nil),
+      textDocument: VersionedTextDocumentIdentifier(uri, version: 0),
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/Clang/ClangLanguageServer.swift:434]: I was wondering about this one as well. `DidChangeTextDocumentNotification` is supposed to take `VersionedTextDocumentIdentifier` and not `OptionalVersionedTextDocumentIdentifier` but apparently clangd has been handling this just fine so far… So maybe you’re right that the cleanest solution is if we just relax our LSP definition here… I still don’t like it though…

**Comment by ahoppen** [comment]
@swift-ci Please test

**Review by benlangmuir — APPROVED** [review]

File: Sources/LanguageServerProtocol/Notifications/WorkDoneProgress.swift [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+public struct WorkDoneProgress: NotificationType {
+  public static var method: String = "$/progress"
+
+  public enum State {
+    case begin(WorkDoneProgressBegin)
+    case report(WorkDoneProgressReport)
+    case end(WorkDoneProgressEnd)
+  }
+
+  public var state: State
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/LanguageServerProtocol/Notifications/WorkDoneProgress.swift:22]: Oh, good idea. I didn’t think of that.

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/LanguageServerProtocol/Error.swift [hunk 9]
```diff
@@ -22,19 +22,68 @@ public struct ErrorCode: RawRepresentable, Codable, Hashable {
...
+  /// A request failed but it was syntactically correct, e.g the
+  /// method name was known and the parameters were valid. The error
+  /// message should contain human readable information about why
+  /// the request failed.
+  public static let RequestFailed = ErrorCode(rawValue: -32803)
+
+  /// The server cancelled the request. This error code should
+  /// only be used for requests that explicitly support being
+  /// server cancellable.
+  public static let ServerCancelled = ErrorCode(rawValue: -32802)
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/LanguageServerProtocol/Error.swift]: I ended up adding a test case: https://github.com/apple/sourcekit-lsp/pull/669/commits/2fe1928d7bc6dd1377110428eae6a72400549530. It appears to work fine.

File: Sources/SourceKitLSP/Clang/ClangLanguageServer.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SourceKitLSP/Clang/ClangLanguageServer.swift:434]: I ended up adding a test case: https://github.com/apple/sourcekit-lsp/pull/669/commits/2fe1928d7bc6dd1377110428eae6a72400549530. It appears to work fine.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Pull request merged**